File: neet_code_all_cpp/math_and_geometry/palindrome_number.cpp
#include <iostream>
#include <string>
#include <vector>

class Solution {
    public:
        // AC
        // int -> std::stringの変換を使う
        bool isPalindrome(int x) {
            if(x < 0) {
                return false;
            }
            std::string x_str = std::to_string(x);
            auto [left, right] = std::pair(0, x_str.size() - 1);
            while(left < right) {
                if(x_str[left] != x_str[right]) {
                    return false;
                }

                ++left;
                --right;
            }

            return true;
        }

        // int -> std::stringを使わないがバッファが要る
        // AC
        int isPalindrome2(int x) {
            if(x < 0) {
                return false;
            } else if(x == 0) {
                return true;
            }

            std::vector<int> digit;
            while(x > 0) {
                digit.push_back(x % 10);
                x /= 10;
            }

            auto [left, right] = std::pair(0, digit.size() - 1);
            while(left < right) {
                if(digit[left] != digit[right]) {
                    return false;
                }

                ++left;
                --right;
            }

            return true;
        }
};

// 模範解答
class SolutionAns {
    public:
        // int -> std::string変換もバッファも使わない
        bool isPalindrome(int x) {
            if(x < 0) {
                return false;
            }

            long div = 1;
            while(x >= 10 * div) {
                div *= 10;
            }

            while(x != 0) {
                int right = x % 10;
                int left = x / div;

                if(left != right) {
                    return false;
                }

                x = (x % div) / 10;
                div /= 100;
            }

            return true;
        }
};

int main(void) {
    int case_1 = 121;
    // => true
    int case_2 = -121;
    // => false
    int case_3 = 10;
    // => false

    Solution s_1;

    std::cout << s_1.isPalindrome(case_1) << std::endl;
    std::cout << s_1.isPalindrome(case_2) << std::endl;
    std::cout << s_1.isPalindrome(case_3) << std::endl;

    std::cout << s_1.isPalindrome2(case_1) << std::endl;
    std::cout << s_1.isPalindrome2(case_2) << std::endl;
    std::cout << s_1.isPalindrome2(case_3) << std::endl;

    SolutionAns s_ans;

    std::cout << s_ans.isPalindrome(case_1) << std::endl;
    std::cout << s_ans.isPalindrome(case_2) << std::endl;
    std::cout << s_ans.isPalindrome(case_3) << std::endl;
}

File: neet_code_all/src/math_and_geometry/palindrome_number.rs
struct Solution;
impl Solution {
    // i32 -> Stringの変換を使う
    // 模範解答と同じ
    pub fn is_palindrome(x: i32) -> bool {
        if x < 0 {
            return false;
        }

        let x = x.to_string().chars().collect::<Vec<char>>();
        let (mut left, mut right) = (0, x.len() - 1);
        while left < right {
            if x[left] != x[right] {
                return false;
            }

            left += 1;
            right -= 1;
        }

        true
    }

    // AC
    // i32 -> Stringの変換を使わないがバッファが必要
    pub fn is_palindrome_2(x: i32) -> bool {
        if x < 0 {
            return false;
        } else if x == 0 {
            return true;
        }

        let mut digigt = vec![];
        let mut x = x;
        while x > 0 {
            digigt.push(x % 10);
            x /= 10;
        }
        let (mut left, mut right) = (0, digigt.len() - 1);
        while left < right {
            if digigt[left] != digigt[right] {
                return false;
            }

            left += 1;
            right -= 1;
        }

        true
    }
}

// C++の模範解答より
// Stringへの変換もバッファも要らない
struct SolutionAnsCpp;
impl SolutionAnsCpp {
    // AC
    pub fn is_palindrome(x: i32) -> bool {
        let mut x = x as i64;
        if x < 0 {
            return false;
        }

        let mut div = 1_i64;
        while x >= 10 * div {
            div *= 10;
        }

        while x != 0 {
            let (right, left) = (x % 10, x / div);

            if left != right {
                return false;
            }

            x = (x % div) / 10;
            div /= 100;
        }

        true
    }
}

fn main() {
    let case_1 = 121;
    // => true
    let case_2 = -121;
    // => false
    let case_3 = 10;
    // => false
    let case_4 = 1410110141;
    // => true

    println!("case_1: {}", Solution::is_palindrome(case_1));
    println!("case_2: {}", Solution::is_palindrome(case_2));
    println!("case_3: {}", Solution::is_palindrome(case_3));

    println!("case_1: {}", Solution::is_palindrome_2(case_1));
    println!("case_2: {}", Solution::is_palindrome_2(case_2));
    println!("case_3: {}", Solution::is_palindrome_2(case_3));

    println!("case_1: {}", SolutionAnsCpp::is_palindrome(case_1));
    println!("case_2: {}", SolutionAnsCpp::is_palindrome(case_2));
    println!("case_3: {}", SolutionAnsCpp::is_palindrome(case_3));
    println!("case_4: {}", SolutionAnsCpp::is_palindrome(case_4));
}
